libr/io/desc.c

```diff
@@ -1,262 +1,216 @@
 	}
-	return desc->plugin->lseek (io, desc, offset, SEEK_SET);
+	return desc->plugin->lseek (desc->io, desc, offset, whence);
 }
 
-R_API void r_io_desc_list (RIO *io) {
-	RIODesc *desc = NULL;
-	RListIter *iter;
-	if (io && io->files) {
-		r_list_foreach (io->files, iter, desc) {
-			if (desc) {
-				io->cb_printf ("- %i", desc->fd);
-				if (desc->uri)
-					io->cb_printf ("\t%s", desc->uri);
-				if (desc->name)
-					io->cb_printf ("\t%s", desc->name);
-				io->cb_printf ("\tstate: %i\tflags: %s\n", desc->state, r_str_rwx_i (desc->flags));
-			}
-		}
+R_API ut64 r_io_desc_size(RIODesc* desc) {
+	ut64 off, ret;
+	if (!desc || !desc->plugin || !desc->plugin->lseek) {
+		return 0LL;
 	}
+	off = desc->plugin->lseek (desc->io, desc, 0LL, R_IO_SEEK_CUR);
+	ret = desc->plugin->lseek (desc->io, desc, 0LL, R_IO_SEEK_END);
+	//what to do if that seek fails?
+	desc->plugin->lseek (desc->io, desc, off, R_IO_SEEK_CUR);
+	return ret;
 }
 
-#if 0
-// XXX: This must be deprecated in order to promote the cast of dataptr to ut32
-R_API int r_io_desc_generate(struct r_io_t *io) {
-	int fd;
-	do fd = 0xf000 + rand ()%0xfff;
-	while (r_io_desc_get(io, fd));
-	return fd;
+R_API bool r_io_desc_exchange(RIO* io, int fd, int fdx) {
+	RIODesc* desc, * descx;
+	SdbListIter* iter;
+	RIOMap* map;
+	if (!(desc = r_io_desc_get (io, fd)) || !(descx = r_io_desc_get (io, fdx)) || !io->maps) {
+		return false;
+	}
+	desc->fd = fdx;
+	descx->fd = fd;
+	r_id_storage_set (io->files, desc,  fdx);
+	r_id_storage_set (io->files, descx, fd);
+	if (io->p_cache) {
+		Sdb* cache = desc->cache;
+		desc->cache = descx->cache;
+		descx->cache = cache;
+		r_io_desc_cache_cleanup (desc);
+		r_io_desc_cache_cleanup (descx);
+	}
+	ls_foreach (io->maps, iter, map) {
+		if (map->fd == fdx) {
+			map->flags &= (desc->flags | R_IO_EXEC);
+		} else if (map->fd == fd) {
+			map->flags &= (descx->flags | R_IO_EXEC);
+		}
+	}
+	return true;
 }
-#endif
-
-R_API void r_io_desc_list_visual(RIO *io, ut64 seek, ut64 len, int width, int use_color) {
-	ut64 mul, min = -1, max = -1;
-	RListIter *iter;
-	RIOMap *s;
-	int j, i;
-
-	width -= 52;
-	if (width<1)
-		width = 30;
 
-	seek = (io->va || io->debug) ? r_io_section_vaddr_to_maddr_try (io, seek) : seek;
+R_API int r_io_desc_get_pid(RIO* io, int fd) {
+	RIODesc* desc;
+	if (!io || !io->files) {
+		//-1 is reserved for plugin internal errors
+		return -2;
+	}
+	if (!(desc = r_io_desc_get (io, fd))) {
+		return -3;
+	}
+	if (!desc->plugin) {
+		return -4;
+	}
+	if (!desc->plugin->isdbg) {
+		return -5;
+	}
+	if (!desc->plugin->getpid) {
+		return -6;
+	}
+	return desc->plugin->getpid (desc);
+}
 
-	r_list_foreach (io->maps, iter, s) {
-		if (min == -1 || s->from < min) {
-			min = s->from;
-		}
-		if (max == -1 || s->to > max) {
-			max = s->to;
-		}
+R_API int r_io_desc_get_tid(RIO* io, int fd) {
+	RIODesc* desc;
+	if (!io || !io->files) {
+		return -2;              //-1 is reserved for plugin internal errors
 	}
-	mul = (max-min) / width;
```

Comment:
I think I put it into the core, maybe we lost it for now. Printing stuff is not a task for io.